### ISSUE TITLE
Utilize maximum CPU threads (7995WX)

### ISSUE THREAD

**Issue by privacyadmin** [NevermindNilas/TheAnimeScripter#30]
Hi,

New user here. I went through the parameters doc and tried to encode a video using the below parameters.

**main.exe --input "E:\example.mkv" --interpolate --interpolate_factor 5 --interpolate_method rife4.17-tensorrt --upscale --upscale_method shufflecugan-tensorrt --upscale_factor 2 --encode_method av1 --nt 192 --buffer_limit 100**

It seems that the nearest parameter that is able to define the number of threads is **--nt** so I went and set it to 192 but it seems that it is not able to do that. 

Is there a way to max out the number of CPU threads to increase TAS performance?

I am using a 7995WX, 768GB RAM and 4090 so I have quite a bit of resources to play around with for TAS

**Comment by privacyadmin** [comment]
To add on, I noticed in the log that SVT might be the limiting factor since it detects 64 cores only

Svt[info]: SVT [version]:       SVT-AV1 Encoder Lib v2.1.0-12-gfcf56491
Svt[info]: SVT [build]  :       GCC 13.2.0       64 bit
Svt[info]: LIB Build date: Jun  2 2024 04:22:55
Svt[info]: -------------------------------------------
Svt[info]: Number of logical cores available: 64
Svt[info]: Number of PPCS 127
Svt[info]: [asm level on system : up to avx512]
Svt[info]: [asm level selected : up to avx512]

**Comment by NevermindNilas** [comment]
Hello,

`--nt` is relatively obsolete due to how multithreadding works in Python, to summarize it is really challenging to bypass the GIL ( Global Interpreter Lock ) in a way in which you can actually benefit from multi core performance. Currently the performance of the TAS is as good as it gets with Python *AFAIK* and getting any more performance out of it would increase the complexity of the script to a point where it would be hard to manage it ( Multiprocessing + Shared Memory Buffers ) and it would likely add additional overhead which would just minimize or even nullify the gains. 

TAS isn't really all that  CPU dependent and most of the logic for decoding and encoding happens on a FFMPEG level which is already heavily multithreaded by nature and all of the data is moved to the GPU as soon and fast as possible so there's not much more room to be gained.

The only potential room for actual large benefits left on the table are memory pinning instead of utilizing paged memory. But that's something I haven't tackled yet.

In a nutshell, parallelization of python is not an easy endeavor and I have other things that are more important to tackle.

For the 2nd comment, that's likely just FFMPEG and u will have to use the `--custom_encoder` feature to play around with FFMPEG's encoding parameters. There's not much more I can do about that.

**Comment by privacyadmin** [comment]
Thanks for sharing your insight. Noted on the complexity and the priorities on hand. I was hoping to eke out as much performance as possible as I am seeing about only about 20% to 40% CPU utilization but do see 90% to 100% of my GPU. Any chance if multi-gpu will be a case in the future?

In regards to SVT-AV1, I recalled Windows has some issue with CPUs with more than 64 threads due to NUMA nodes. When the chance arises, I shall test TAS in Linux environment to see if this can be overcome.

Thank you for sharing TAS and your support. I shall close this ticket now.

**Comment by NevermindNilas** [comment]
Multi GPU is possible but yet again it is such an edge case scenario that it would likely be on the bottom of the list for support.

Currently my main priorities are:

Proper Linux support ( GUI & CLI )
MacOS support ( will need a MacBook for that so likely impossible without the support from github sponsors )
Flow extraction for after effects 
10 bit encoding.

Anyhow, thank you for using TAS and do let me know if you have any other issues. Also feel free to join the discord server: https://discord.com/invite/GVHrTQRMNq
